### ISSUE TITLE
Can't use Default property on a control array

### ISSUE THREAD

**Issue by dmrvb** [twinbasic/twinbasic#1240]
**Describe the bug**

Can't set default proprty on a control array, get error "This property is read only"

**To Reproduce**
Put a textbox TBa on a form.
Put an array of textboxes TBb(1) on a form.

TBa = "mytext" works ok
TBb(1)="mytext" gives the error

**Expected behavior**
Default property needs to work to be compatible with VB.   but see below

**Desktop (please complete the following information):**
 - OS: [e.g. Windows 11]
 - twinBASIC compiler version [e.g. 145]

I don't like using the default property, its error prone, I only found the above bug because I had forgot to specify .Text and VB gave no error but TB did.
Could/should TB give a warning on default/unspecified control properties???. This would be logical as it already  gives a variant warning on unspecified data types wchich is related lazy/poor programming.

**Comment by WaynePhillipsEA** [comment]
> Could/should TB give a warning on default/unspecified control properties???. This would be logical as it already gives a variant warning on unspecified data types wchich is related lazy/poor programming.

I do agree.  My only concern is that some imported projects might light up like a Xmas tree.   But the same argument could be made against the 'defaults to Variant' warning that we already have in place.

Perhaps to help reduce noise, we need a filter in the `Problems` pane so that we can turn off **warnings** from appearing in the Problems list, and by default just show **errors** there.    The warnings would still be underlined in the code editor of course.

**Comment by dmrvb** [comment]
In the settings/compiler warnings section you already give the option for error/warning/ignore on various code issues, which is really good, can you do the same with default control properties?

**Comment by WaynePhillipsEA** [comment]
Yes, all warnings get that option automatically.   My concern is in regards to a newcomer opening their VB6 project in tB and being immediately overwhelmed with masses of "warnings".   I may just be overthinking it.   Don't get me wrong; I absolutely agree with your suggestion and think we should offer it.

**Comment by dmrvb** [comment]
You are absolutely correct, the new user first impression is vital. If its not positive then its very hard to fix that later.
A new tB installation could start up in a new user mode with options suitably set.  The first import of a VB project should ideally work without any errors. How about a message box then pops up saying "tB can compile this but there are XXX code weaknesses that could be improved", and an option to turn on the error checkers and an "option strict".

**Comment by wqweto** [comment]
> Perhaps to help reduce noise, we need a filter in the `Problems` pane so that we can (optionally) turn off **warnings** from appearing in the Problems list, and by default just show **errors** there.

And then probably add `Warnings as errors` option to "expand" into the other direction where (sane) teams would like to have zero warnings source codebase.

Of course best would be a separate linter as some teams don't allow `On Error Resume Next` and similar straight-forward cowboy-style one-off programming practices.

**Comment by dmrvb** [comment]
'On Error Resume Next' can be a really useful technique as long as it is used thoughtfully and not just as a midless "catch unknown errors". I don't know why it gets so much criticism, I think its just one of those things that the VB bashers have latched on to.
I would be happy for it to be a warning, but would prefer it not to be banned by an "Option Strict".

This is an elegant bit of code????:........
On Error Resume Next
For Each F in Forms
  call "method that only some forms have"
Next

Its also good when trying to recover from a partial crash, calling clean up routines when the state of things is unknown.
(and the part crash might be due to a Windows failure rather than my code :-)
And whilst Im ranting, GOTO is also pretty good if used with care, sometimes more readable than convoluted nested IFs.

**Comment by WaynePhillipsEA** [comment]
Perhaps we don't even need an `Option Strict` mode anyway?  If all warnings from the compiler can be promoted to errors on an individual basis, then you can pick and choose which ones you want to adhere to.  So we might decide to offer a hint-level warning for use of `On Error Resume Next`, and then if you want to prevent use of it, set it Error rather than the default of Ignore.

**Comment by wqweto** [comment]
> And whilst Im ranting, GOTO is also pretty good if used with care, sometimes more readable than convoluted nested IFs.

GoTo and OERN are not in the same league provided that GoTo is used extensively for error handling in any C codebase incl. Linux kernel.

On the other hand OERN is pure evil e.g. in your sample code you potentially won't be able to log the Out Of Memory error which `Forms` property raises on an obscure client machine in the country.

The root cause for using OERN in your sample is the inability of the language to call "method that only some forms have" without raising an error just for this line.

In our codebases we use a [`CallByName` replacement function that never raises](function ) like `Call DispInvoke(F, "method that only some forms have")` exactly for the purpose of never having an excuse to use OERN ever.

**Comment by dmrvb** [comment]
> Perhaps we don't even need an `Option Strict` mode anyway? If all warnings from the compiler can be promoted to errors on an individual basis, then you can pick and choose which ones you want to adhere to. So we might decide to offer a hint-level warning for use of `On Error Resume Next`, and then if you want to prevent use of it, set it Error rather than the default of Ignore.
''''''''''''''''''''''''''''''''
Im not too bothered, maybe others with stronger feelings on this matter should decide. I like the idea of "Option Strict" as its a single elegant statement to force us to write better code, but then again I like the Python philosophy that there should be only one obvious way to do things, and setting individual error/warnings is certainly the more powerful and flexible way of doing things. So, yes, I don't need Option Strict.

**Comment by dmrvb** [comment]
On Error Resume Next

I have certainly accidentally used this badly and created hard to find bugs where errors that I should be reporting get ignored, but then any language construct can be used badly.
My other big use of OERN is to do a last ditch clean up after a partial crash. I want to get things at least half working so that I can report something to the user and get things stable enough so that the user can save their data.
Something that I did not anticipate or expect to happen has gone wrong so running clean up routines with an OERN is the best aproach. My personal rule is that OERN should only be used to protect a small and well defined section of code.

**Comment by fafalone** [comment]
There are some APIs that throw a runtime error instead of returning an error code for normal, common situation where something is not available or the user cancelled; I think OERN is perfectly acceptable there because it should have just been a normal 'Nope, not available' return code for which it's a common outcome; you wouldn't want to log it through a regular error handler because it's not something avoidable or fixable. Then it's often the case you have to validate the results *anyway* because there's other reasons the object might be null or whatever.

So yeah, there are a few circumstances where OERN is perfectly reasonable for single lines or small blocks of code. I think some people just don't want to be associated with something so widely misused.

**Comment by bclothier** [comment]
Just to posit a possible different approach. Instead of warnings (or in conjunction with warnings), offer a CodeLens feature where the default member are spelled out in light grey text to make it more obvious what it's really doing. 

RE: OERN. I think what we really want is an operator similar to null-safe operator:

```
For Each c in .Controls
  c?.SomeMemberThatExistsOnOnlyCertainTypes = "foo"
Next
```

That would eliminate this common use for OERN and make for more terse/readable code that doesn't get bloated with all error handling logic.

**Comment by dmrvb** [comment]
We are slightly mixing two topics here, one is what to do about OERN, the other is language changes to make OERN redundant.
I propose that we just have to live with OERN because its already part of VB,
As to language enhancements to reduce the need for OERN then yes, lets have them, first get tB compatible with VB, then make it better.

Dim T as textbox
For Each T in Controls
...
Next
This will fail if the controls collection contains anything other than textboxes, even if that control is not referenced.
OERN is the easiest way to handle this.
The other way:
Dim C as control
for each C in controls
  if Typeof C is textbox
  
End if
next 
always feels a bit clumsy to me.

**Comment by wqweto** [comment]
Btw, we've been using `Break on All Errors` option on all our IDE installations for many (10+) years now so life without OERN **is** possible :-))

In this regard OERN does not silence any error when `Break on All Errors` is set -- it's just not possible.

With this option set upon error the `Err` object does not get populated when the IDE breaks. The interpreter just stops at the moment just before raising the error and populating `Err`, before jumping to error handler label or "swallowing" everything if OERN-ing.

The idea is that error handling code, the one after `ProcError` label, should never be executed in the IDE. Its main purpose is to log **unexpected** at time of development circumstances so that we can handle these with sane user-facing messages or some other retrying or reconciliatory logic.

This is like swallowing Unhandled exceptions in C# with no logging or showing message boxes with exception description to the clueless end-user. None of it makes sense.

**Comment by WaynePhillipsEA** [comment]
Fixed in BETA 146.  Thanks for reporting :)
